### ISSUE TITLE
mopidy-tidal and mopidy 3.0, python 3.7

### ISSUE THREAD

**Issue by tehkillerbee** [mopidy/mopidy#1866]
Hello everyone,
After upgrading to the latest version of mopidy 3.0, I discovered that the unofficial(?) mopidy-tidal extension is not working anymore. Furthermore, the original author has archived the project.

However, I wanted to continue mopidy and tidal together. I discovered a few smaller bugs and have fixed them in my mopidy-tidal fork available here https://github.com/tehkillerbee/mopidy-tidal

Is there a reason the mopidy-tidal extension is still not officially supported?

**Comment by adamcik** [comment]
Feel free to send a PR adding it to https://github.com/mopidy/website once it's in a usable state and installable via pip. As for "official" support there isn't really such a thing in the mopidy ecosystem, closest we have is that some extensions live in the Mopidy github org, but there isn't really any benefit to that other than admins of the mopidy org having access (at least as I see it).

**Comment by kingosticks** [comment]
Just be warned, Tidal don't share their API details unless you give them a load of money. The API key used by this extension is 'borrowed' from somewhere else, I'm surprised it still works!

Can we close this since there's nothing actionable for us here?

**Comment by jodal** [comment]
@tehkillerbee I'd recommend getting in contact with the original Mopidy-Tidal author to see if you can get access to the PyPI project so that you can upload new releases under the same name.

**Comment by johnhamelink** [comment]
Hey @tehkillerbee, I've been trying out your fork and it works very well! I have found one bug which I triggered when I tried to view the `[My Tracks]` playlist inside `Browse -> Tidal`:

```
INFO     2020-01-31 11:09:11,988 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:3581763:88596710:88596738'
ERROR    2020-01-31 11:09:12,726 [363746:Core-7] mopidy.core.library
  TidalBackend backend caused an exception.
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 17, in _backend_error_handling
    yield
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 217, in lookup
    result = future.get()
  File "/usr/lib/python3.8/site-packages/pykka/_threading.py", line 45, in get
    _compat.reraise(*self._data['exc_info'])
  File "/usr/lib/python3.8/site-packages/pykka/_compat/__init__.py", line 29, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 193, in _actor_loop
    response = self._handle_receive(envelope.message)
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 299, in _handle_receive
    return callee(*message.args, **message.kwargs)
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/library.py", line 191, in lookup
    tracks += self._lookup_track(session, parts)
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/library.py", line 205, in _lookup_track
    self.lru_album_tracks[album_id] = tracks
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/lru_cache.py", line 35, in __setitem__
    self._check_limit()
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/lru_cache.py", line 40, in _check_limit
    k = self.keys()[0]
TypeError: 'odict_keys' object is not subscriptable
```

I suspect this might be an API backpressure issue as the first few requests go through successfully:

```
INFO     2020-01-31 11:08:57,098 [363746:TidalBackend-6] mopidy_tidal.library
  Browsing uri tidal:directory
INFO     2020-01-31 11:08:57,498 [363746:TidalBackend-6] mopidy_tidal.library
  Browsing uri tidal:my_tracks
INFO     2020-01-31 11:08:58,300 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:6038062:109936827:109936828'
INFO     2020-01-31 11:08:58,730 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:08:58,733 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:9715341:105179343:105179344'
INFO     2020-01-31 11:08:59,168 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:08:59,170 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:4709193:98228092:98228093'
INFO     2020-01-31 11:08:59,615 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:08:59,616 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:1003:90502209:90502217'
INFO     2020-01-31 11:09:00,064 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:00,066 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:7165490:104721842:104721843'
INFO     2020-01-31 11:09:00,490 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:00,493 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:653:39817:39829'
INFO     2020-01-31 11:09:00,951 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:00,953 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:10833260:102342354:102342356'
INFO     2020-01-31 11:09:01,380 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:01,382 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:8010954:114594764:114594765'
INFO     2020-01-31 11:09:01,808 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:01,810 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:8240136:85780817:85780819'
INFO     2020-01-31 11:09:02,339 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:02,341 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:5207648:94897179:94897180'
INFO     2020-01-31 11:09:02,502 [363746:MpdSession-18] mopidy_mpd.session
  New MPD connection from [::1]:46730
INFO     2020-01-31 11:09:02,770 [363746:TidalBackend-6] mopidy_tidal.library
  Returning 1 tracks
INFO     2020-01-31 11:09:02,774 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:3519582:10631854:10631857'
ERROR    2020-01-31 11:09:03,216 [363746:Core-7] mopidy.core.library
  TidalBackend backend caused an exception.
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 17, in _backend_error_handling
    yield
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 217, in lookup
    result = future.get()
  File "/usr/lib/python3.8/site-packages/pykka/_threading.py", line 45, in get
    _compat.reraise(*self._data['exc_info'])
  File "/usr/lib/python3.8/site-packages/pykka/_compat/__init__.py", line 29, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 193, in _actor_loop
    response = self._handle_receive(envelope.message)
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 299, in _handle_receive
    return callee(*message.args, **message.kwargs)
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/library.py", line 191, in lookup
    tracks += self._lookup_track(session, parts)
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/library.py", line 205, in _lookup_track
    self.lru_album_tracks[album_id] = tracks
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/lru_cache.py", line 35, in __setitem__
    self._check_limit()
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/lru_cache.py", line 40, in _check_limit
    k = self.keys()[0]
TypeError: 'odict_keys' object is not subscriptable
INFO     2020-01-31 11:09:03,222 [363746:TidalBackend-6] mopidy_tidal.library
  Lookup uris 'tidal:track:3736815:97122857:97122860'
ERROR    2020-01-31 11:09:03,655 [363746:Core-7] mopidy.core.library
  TidalBackend backend caused an exception.
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 17, in _backend_error_handling
    yield
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 217, in lookup
    result = future.get()
  File "/usr/lib/python3.8/site-packages/pykka/_threading.py", line 45, in get
    _compat.reraise(*self._data['exc_info'])
  File "/usr/lib/python3.8/site-packages/pykka/_compat/__init__.py", line 29, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 193, in _actor_loop
    response = self._handle_receive(envelope.message)
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 299, in _handle_receive
    return callee(*message.args, **message.kwargs)
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/library.py", line 191, in lookup
    tracks += self._lookup_track(session, parts)
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/library.py", line 205, in _lookup_track
    self.lru_album_tracks[album_id] = tracks
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/lru_cache.py", line 35, in __setitem__
    self._check_limit()
  File "/usr/lib/python3.8/site-packages/Mopidy_Tidal-0.2.2-py3.8.egg/mopidy_tidal/lru_cache.py", line 40, in _check_limit
    k = self.keys()[0]
TypeError: 'odict_keys' object is not subscriptable
```

I hope this is helpful information :)

**Comment by tehkillerbee** [comment]
Thanks @johnhamelink! I have managed to replicate that issue as well when using the default web interface. It looks like another python3 vs. python2 incompatibility that I did not catch when I first tested the library. Interestingly, it does not occur when using the Mopidy-Mobile web extension I am currently using.

I think it will be necessary to verify/test that the Mopidy Tidal plugin does not rely on any other unsupported python 2 code. Maybe it will be necessary to write a simple test script with good enough coverage... Feel free to give any suggestions!

I will try to get hold of the original author and see if we can get the original extension updated on PyPI
